src/device/wg_controller.rs

```diff
@@ -516,7 +516,18 @@ async fn select_endpoint_for_peer<'a>(
                 public_key,
                 upgrade_request_endpoint
             );
-            Ok((Some(*upgrade_request_endpoint), None))
+            if let Some(peer) = actual_peer {
+                telio_log_debug!("Our actual endpoint is: {:?}", peer.endpoint);
+            }
+
+            if (peer_state == PeerState::Upgrading || peer_state == PeerState::Direct)
+                && actual_endpoint.is_some()
+                && actual_endpoint != *proxy_endpoint
+            {
+                Ok((actual_endpoint, None))
```

Comment:
Please add the `telio_log_debug!()` statements in such a way, that it would be clear whether we are keeping `actual_endpoint` or using the `upgrade_request_endpoint`.